### PR TITLE
UniGo66 keyboard added

### PR DESCRIPTION
**Pull request by danielhklein**
## Types of changes

- [ ] Core
- [ ] Bugfix
- [ ] New Feature
- [ ] Enhancement/Optimization
- [x] Keyboard (addition or update)
- [x] Keymap/Layout/Userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by this PR

* 

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/sirius/unigo66/rules.mk [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# MCU name
+MCU = atmega32u4
+
+
+F_CPU ?= 16000000
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/rules.mk]: Should just be an `=` here.

```suggestion
F_CPU = 16000000
```

File: keyboards/sirius/unigo66/info.json [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+				{"label":"O", "x":13.5, "y":1},
+				{"label":"P", "x":14.5, "y":1},
+				{"label":"\\", "x":15.5, "y":1, "w":1.5},
+				{"label":"Caps", "x":0, "y":2, "w":1.5},
+				{"label":"A", "x":1.5, "y":2},
+				{"label":"S", "x":2.5, "y":2},
+				{"label":"D", "x":3.5, "y":2},
+				{"label":"F", "x":4.5, "y":2},
+				{"label":"G", "x":5.5, "y":2},
+				{"label":"L2", "x":6.5, "y":2.5, "h":1.5},
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/info.json]: The order of JSON objects doesn't match the order in `unigo66.h`, which will lead to misplaced keys if someone uses QMK Configurator to build their firmware.

There's a series of changes needed to fix this file.

```diff
diff --git a/keyboards/sirius/unigo66/info.json b/keyboards/sirius/unigo66/info.json
index 3a38bea74..250a87b63 100644
--- a/keyboards/sirius/unigo66/info.json
+++ b/keyboards/sirius/unigo66/info.json
@@ -41,8 +41,6 @@
 				{"label":"D", "x":3.5, "y":2},
 				{"label":"F", "x":4.5, "y":2},
 				{"label":"G", "x":5.5, "y":2},
-				{"label":"L2", "x":6.5, "y":2.5, "h":1.5},
-				{"label":"]", "x":9.5, "y":2.5, "h":1.5},
 				{"label":"H", "x":10.5, "y":2},
 				{"label":"J", "x":11.5, "y":2},
 				{"label":"K", "x":12.5, "y":2},
@@ -55,6 +53,8 @@
 				{"label":"C", "x":3.5, "y":3},
 				{"label":"V", "x":4.5, "y":3},
 				{"label":"B", "x":5.5, "y":3},
+				{"label":"L2", "x":6.5, "y":2.5, "h":1.5},
+				{"label":"]", "x":9.5, "y":2.5, "h":1.5},
 				{"label":"N", "x":10.5, "y":3},
 				{"label":"M", "x":11.5, "y":3},
 				{"label":",", "x":12.5, "y":3},
@@ -68,12 +68,12 @@
 				{"label":"Down", "x":14.5, "y":4},
 				{"label":"Right", "x":15.5, "y":4},
 				{"label":"Space1", "x":5.5, "y":5},
-				{"label":"Space2", "x":5.5, "y":6},
 				{"label":"PgUp", "x":6.5, "y":5},
-				{"label":"PgDn", "x":6.5, "y":6},
 				{"label":"Home", "x":9.5, "y":5},
-				{"label":"End", "x":9.5, "y":6},
 				{"label":"Enter1", "x":10.5, "y":5},
+				{"label":"Space2", "x":5.5, "y":6},
+				{"label":"PgDn", "x":6.5, "y":6},
+				{"label":"End", "x":9.5, "y":6},
 				{"label":"Enter2", "x":10.5, "y":6}
 			]
 		}
```

File: keyboards/sirius/unigo66/keymaps/danielhklein/keymap.c [hunk 3]
```diff
@@ -0,0 +1,189 @@
+#include "unigo66.h"
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/keymaps/danielhklein/keymap.c]: This should be

```suggestion
#include QMK_KEYBOARD_H
```

File: keyboards/sirius/unigo66/keymaps/danielhklein/keymap.c [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  KC_F7,   KC_F8,  KC_F9,   KC_F10,  KC_F11,  KC_F12,  XXXXXXX, XXXXXXX, KC_P7,   KC_P8,   KC_P9,   KC_PPLS, XXXXXXX, XXXXXXX,
+  XXXXXXX, XXXXXXX,XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   KC_P4,   KC_P5,   KC_P6,   KC_PENT, XXXXXXX, XXXXXXX, 
+  XXXXXXX, XXXXXXX,XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, KC_P1,   KC_P2,   KC_P3,   KC_SPC,  XXXXXXX, XXXXXXX,
+  XXXXXXX, XXXXXXX,XXXXXXX,                                                                         KC_P0,   KC_PDOT, XXXXXXX,
+                                              XXXXXXX, XXXXXXX, _______, XXXXXXX,
+                                              XXXXXXX, XXXXXXX, _______, XXXXXXX
+    )
+};
+
+const uint16_t PROGMEM fn_actions[] = {
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/keymaps/danielhklein/keymap.c]: Lines 133-142 aren't needed at all and should be deleted. Both functions are deprecated.

File: keyboards/sirius/unigo66/keymaps/default/config.h [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#ifndef CONFIG_USER_H
+#define CONFIG_USER_H
+
+#include "../../config.h"
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/keymaps/default/config.h]: Line 4 should be updated as below. QMK's build script automatically includes all the `config.h` files in the build tree, so this line is redundant. There should be a note, though, for users who copy the keymap to know where the overrides go.

```suggestion
// place overrides here
```

File: keyboards/sirius/unigo66/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
+#include "unigo66.h"
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/keymaps/default/keymap.c]: This should be

```suggestion
#include QMK_KEYBOARD_H
```

File: keyboards/sirius/unigo66/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	______,  ______,  ______,  ______,  ______,  ______,  ______,    ______,  ______,  ______,  ______,  ______,  ______,  ______, 
+	______,  ______,  ______,  ______,  ______,  ______,                      ______,  ______,  ______,  ______,  ______,  ______, 
+	______,  ______,  ______,  ______,  ______,  ______,  ______,    ______,  ______,  ______,  ______,  ______,  ______,  ______, 
+	______,  ______,  ______,                                                                            ______,  ______,  ______,
+                                               ______,  ______,    ______,  ______,  
+                                               ______,  ______,    ______,  ______
+    ),
+};
+
+const uint16_t PROGMEM fn_actions[] = {
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/keymaps/default/keymap.c]: Lines 52-61 are unused and should be deleted, as these functions are deprecated.

File: keyboards/sirius/unigo66/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	KC_ESC,  KC_1,   KC_2,   KC_3,   KC_4,   KC_5,   KC_MINS,       KC_EQL,  KC_6,   KC_7,   KC_8,   KC_9,   KC_0,    KC_BSPC,
+	KC_TAB,  KC_Q,   KC_W,   KC_E,   KC_R,   KC_T,   TG(1),         KC_LBRC, KC_Y,   KC_U,   KC_I,   KC_O,   KC_P,    KC_BSLS,
+	KC_CAPS, KC_A,   KC_S,   KC_D,   KC_F,   KC_G,                           KC_H,   KC_J,   KC_K,   KC_L,   KC_SCLN, KC_QUOT, 
+	KC_LSFT, KC_Z,   KC_X,   KC_C,   KC_V,   KC_B,   TG(2),         KC_RBRC, KC_N,   KC_M,   KC_COMM,KC_DOT, KC_SLSH, KC_UP  ,
+	KC_LCTL, KC_LGUI,KC_LALT,                                                                        KC_LEFT,KC_DOWN, KC_RGHT,
+                                           KC_SPC, KC_PGUP,       KC_HOME, KC_ENT,
+                                           KC_SPC, KC_PGDN,       KC_END,  KC_ENT
+    ),
+    [1] = LAYOUT(
+  ______,  ______,  ______,  ______,  ______,  ______,  ______,    ______,  ______,  ______,  ______,  ______,  ______,  ______,
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/keymaps/default/keymap.c]: Every instance of `______` from Lines 33-48 should be replaced with `_______`.

(more on this below)

File: keyboards/sirius/unigo66/unigo66.h [hunk 9]
```diff
@@ -0,0 +1,105 @@
+#ifndef UNIGO66_H
+#define UNIGO66_H
+
+#include "quantum.h"
+
+#define ______ KC_TRNS
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/unigo66.h]: QMK defines a keycode for this automatically.

https://github.com/qmk/qmk_firmware/blob/a0309db983150fd0197eb620a6ba552d90d29f93/quantum/quantum_keycodes.h#L29-L31

```suggestion
```

Hence the big update for Lines 33-48 of the default `keymap.c`.

File: keyboards/sirius/unigo66/unigo66.h [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+ * `--------+------+------+---------------------------'           `---------------------------+------+------+--------'
+ *   | Ctrl | Super|  Alt |                                                                   | Left |  Down| Right |
+ *   `--------------------'                                                                   `--------------------'
+ *                                        ,------|------.       ,---------------.
+ *                                        |      | PgUp |       | Home |        |
+ *                                        | Space|------|       |------|  Enter |
+ *                                        |      | PgDn |       | End  |        |
+ *                                        `-------------'       `---------------'
+ */
+#define LAYOUT( \
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/unigo66.h]: Better visual alignment:

```diff
diff --git a/keyboards/sirius/unigo66/unigo66.h b/keyboards/sirius/unigo66/unigo66.h
index 98798adf8..3215c891b 100644
--- a/keyboards/sirius/unigo66/unigo66.h
+++ b/keyboards/sirius/unigo66/unigo66.h
@@ -88,11 +86,11 @@
 #define LAYOUT( \
   K29,K1E,K1F,K20,K21,K22,K2D,    K2E,K23,K24,K25,K26,K27,K2A,\
   K2B,K14,K1A,K08,K15,K17,K4B,    K2F,K1C,K18,K0C,K12,K13,K31,\
-  K39,K04,K16,K07,K09,K0A,        K0B,K0D,K0E,K0F,K33,K34,\
+  K39,K04,K16,K07,K09,K0A,            K0B,K0D,K0E,K0F,K33,K34,\
   KE1,K1D,K1B,K06,K19,K05,K4E,    K30,K11,K10,K36,K37,K52,KE5,\
-  KE0,KE3,KE2,            K50,K51,K4F,\
-                K2C,K49,    K4A,K28,        \
-                KE6,K4C,    K4D,KE4         \
+  KE0,KE3,KE2,                                    K50,K51,K4F,\
+                      K2C,K49,    K4A,K28,        \
+                      KE6,K4C,    K4D,KE4         \
 ) LAYOUT_ALL( \
             KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,                                              \
     K29,    KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,      KC_NO,KC_NO,KC_NO,  KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,\
```

File: keyboards/sirius/unigo66/README.md [hunk 11]
```diff
@@ -0,0 +1,17 @@
+# UniGo66
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/README.md]: This file should be renamed `readme.md` (fully lowercase).

File: keyboards/sirius/unigo66/keymaps/danielhklein/config.h [hunk 12]
```diff
@@ -0,0 +1,6 @@
+#ifndef CONFIG_USER_H
+#define CONFIG_USER_H
+
+#include "../../config.h"
```
Comment (noroadsleft) [inline keyboards/sirius/unigo66/keymaps/danielhklein/config.h]: Line 4 should be updated as below. QMK's build script automatically includes all the `config.h` files in the build tree, so this line is redundant. There should be a note, though, for users who copy the keymap to know where the overrides go.

```suggestion
// place overrides here
```

**Comment by noroadsleft** [comment]
Putting this here to get more eyes on it because I'm a bit out of my depth here.

@danielhklein sent me [this Gist](https://gist.github.com/danielhklein/5dac736be5f799bac96f69dbde9a18f3) via Discord on Tuesday, saying, in part:

> Hey, on the PR for the UniGo66, adding the #include "config_common.h" breaks build, I'm not 100% sure why but I have suspicions
> I think it's because of the weird custom_matrix implementation, but honestly I wouldn't know where to begin in debugging or rewriting that

I know that the error is because one file is trying to define variables that are already defined, specifically these:

https://github.com/qmk/qmk_firmware/blob/0ba9620d6541e5ade89271eccc7371d5a8a3ad15/tmk_core/protocol/usb_hid/arduino-1.0.1/cores/arduino/binary.h#L4
https://github.com/qmk/qmk_firmware/blob/0ba9620d6541e5ade89271eccc7371d5a8a3ad15/tmk_core/protocol/usb_hid/arduino-1.0.1/cores/arduino/binary.h#L12

which are already defined here:

https://github.com/qmk/qmk_firmware/blob/0ba9620d6541e5ade89271eccc7371d5a8a3ad15/quantum/config_common.h#L80-L81

I don't know quite enough to say whether including `config_common.h` is required for every build, and as I don't have the keyboard in question, I can't test it.

**Comment by danielhklein** [comment]
@noroadsleft thanks for posting that here. I actually have some updated files for the config and the matrix, I will be testing them tonight but I suspect they may resolve these issues. So before anyone dives too deep, just wanted to say that this may already be solved.

**Comment by danielhklein** [comment]
All suggested changes have been accepted other than the inclusion of `#include config_common.h`. in discussion with @drashna on Discord, this should be okay due to the configuration of UniGo66

File: keyboards/sirius/unigo66/config.h [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+/* size of virtual matrix */
+#define MATRIX_ROWS 16
+#define MATRIX_COLS 16
+
+/* matrix scanning is done in custom_matrix.cpp */
+#define DIODE_DIRECTION CUSTOM_MATRIX
+
+/* key combination for command */ 
+#define IS_COMMAND() (keyboard_report->mods == (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT)))
```
Comment (vomindoraan) [inline keyboards/sirius/unigo66/config.h]:
```suggestion
```
This is no longer required. Please remove lines 38–40 (inclusive) in this file.

Comment (vomindoraan) [reply keyboards/sirius/unigo66/config.h]: Thanks!

**Comment by danielhklein** [comment]
odd, I've removed it from my repo. It hasn't pushed to this request though. Looking into how to handle that...

**Comment by danielhklein** [comment]
@noroadsleft I don't think the keyboards/unigo66 files still exist. They don't show up in my repo at latest commit, and from the PR's "Files Changed" section they also are not present

**Comment by drashna** [comment]
Thanks!

**Pull request merged**